include/llvm/Support/WinTypes.h

```diff
@@ -77,376 +140,454 @@ typedef const GUID &REFGUID;
+} STATSTG;
+
+enum tagSTATFLAG {
+  STATFLAG_DEFAULT = 0,
+  STATFLAG_NONAME = 1,
+  STATFLAG_NOOPEN = 2
+};
+
+//===--------------------- UUID Related Macros ----------------------------===//
+
 // The following macros are defined to facilitate the lack of 'uuid' on Linux.
 #define DECLARE_CROSS_PLATFORM_UUIDOF(T)                                       \
 public:                                                                        \
   static REFIID uuidof() { return static_cast<REFIID>(&T##_ID); }              \
+                                                                               \
 private:                                                                       \
-  static const char T##_ID;                                                    \
+  static const char T##_ID;
 
 #define DEFINE_CROSS_PLATFORM_UUIDOF(T) const char T::T##_ID = '\0';
 #define __uuidof(T) T::uuidof()
-#define IID_PPV_ARGS(ppType) (**(ppType)).uuidof(), reinterpret_cast<void**>(ppType)
-
-#define interface struct
+#define IID_PPV_ARGS(ppType)                                                   \
+  (**(ppType)).uuidof(), reinterpret_cast<void **>(ppType)
 
-class CAllocator {
-public:
-  static void* Reallocate(void* p, size_t nBytes) throw() {
-    return realloc(p, nBytes);
-  }
-  static void* Allocate(size_t nBytes) throw() {
-    return malloc(nBytes);
-  }
-  static void Free(void* p) throw() {
-    free(p);
-  }
-};
+//===--------------------- COM Interfaces ---------------------------------===//
 
 // Note: This implementation is inspired by:
 // https://msdn.microsoft.com/en-us/library/office/cc839627.aspx
-interface IUnknown {
-  virtual HRESULT QueryInterface(REFIID riid, void** ppvObject) = 0;
+struct IUnknown {
+  virtual HRESULT QueryInterface(REFIID riid, void **ppvObject) = 0;
   virtual ULONG AddRef() {
     ++m_count;
     return m_count;
   }
   virtual ULONG Release() {
     --m_count;
-    if(m_count == 0) {
+    if (m_count == 0) {
       delete this;
     }
     return m_count;
   }
   virtual ~IUnknown() {}
 
-  template <class Q>
-  HRESULT QueryInterface(Q** pp) {
-    return QueryInterface(__uuidof(Q), (void**)pp);
+  template <class Q> HRESULT QueryInterface(Q **pp) {
+    return QueryInterface(__uuidof(Q), (void **)pp);
   }
 
-  private:
+private:
   std::atomic<unsigned long> m_count;
 
-  DECLARE_CROSS_PLATFORM_UUIDOF(IUnknown);
+  DECLARE_CROSS_PLATFORM_UUIDOF(IUnknown)
 };
 
 struct INoMarshal : public IUnknown {
-  DECLARE_CROSS_PLATFORM_UUIDOF(INoMarshal);
+  DECLARE_CROSS_PLATFORM_UUIDOF(INoMarshal)
 };
 
-template <class T>
-class CComPtrBase {
+struct IMalloc : public IUnknown {
+  virtual void *Alloc(size_t size) { return malloc(size); }
+  virtual void *Realloc(void *ptr, size_t size) { return realloc(ptr, size); }
+  virtual void Free(void *ptr) { free(ptr); }
+  virtual HRESULT QueryInterface(REFIID riid, void **ppvObject) {
+    assert(false && "QueryInterface not implemented for IMalloc.");
+    return -1;
```

Comment:
It probably won't have any effect, but MSDN says that QueryInterface returns E_NOINTERFACE when the interface isn't supported. I guess that's not exactly what's happening here though.